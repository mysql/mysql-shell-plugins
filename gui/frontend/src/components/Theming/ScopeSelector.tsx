/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";
import {
    Component, IComponentState, IComponentProperties, Popup, ComponentPlacement, Label, TagInput,
    Orientation, Button, ITag,
} from "../ui";
import { ValueEditDialog, IDialogValues, IDialogValidations, IDialogSection } from "../Dialogs/ValueEditDialog";
import { ThemeEditorLists } from "./ThemeEditorLists";
import { DialogResponseClosure } from "../../app-logic/Types";

export interface IScopeSelectorProperties extends IComponentProperties {
    defaultScopes: string[];
    customScopes: string[];
}

interface IScopeSelectorState extends IComponentState {
    defaultTags: ITag[];           // Tags derived from props.defaultScopes.
    currentCustomScopes: string[]; // Custom tags including user-added entries.
}

// A popup that shows lists of standard and custom scopes as tags, which can be used to assign any of them
// to a syntax color token entry in the theme editor.
export class ScopeSelector extends Component<IScopeSelectorProperties, IScopeSelectorState> {

    private newScopeDialogRef = React.createRef<ValueEditDialog>();
    private popupRef = React.createRef<Popup>();

    public constructor(props: IScopeSelectorProperties) {
        super(props);

        this.state = {
            defaultTags: props.defaultScopes.map((tag: string) => {
                return { id: tag, caption: tag };
            }),
            currentCustomScopes: props.customScopes,
        };
    }

    public componentDidUpdate(prevProps: IScopeSelectorProperties): void {
        const { customScopes } = this.props;
        if (customScopes !== prevProps.customScopes) {
            this.setState({ currentCustomScopes: customScopes });
        }
    }

    public render(): React.ReactNode {
        const { defaultTags, currentCustomScopes } = this.state;

        const customTags: ITag[] = [];
        currentCustomScopes.forEach((tag: string) => {
            if (!ThemeEditorLists.defaultScopes.includes(tag)) {
                customTags.push({
                    id: tag,
                    caption: tag,
                });
            }
        });


        return (
            <Popup
                ref={this.popupRef}
                placement={ComponentPlacement.RightCenter}
                className="scopeSelector"
            >
                <Label caption="Scope Selector" id="title" />
                <Label
                    caption={
                        "Use the collections to assign scopes to each color entry, by dragging a scope tag" +
                        " with the mouse into the associated scope field."
                    }
                    id="description"
                />
                <Label caption="Default Scopes" id="subTitle" />
                <TagInput
                    tags={defaultTags}
                    orientation={Orientation.TopDown}
                />
                <Label caption="Defined Scopes" id="subTitle" />
                <Label
                    caption={
                        "Unused scopes are automatically removed on next load of the theme, so make sure you" +
                        " use a scope at least once to keep it in this list."
                    }
                    id="description"
                />
                <TagInput
                    tags={customTags}
                    orientation={Orientation.TopDown}
                />
                <div className="footer verticalCenterContent">
                    <ValueEditDialog
                        ref={this.newScopeDialogRef}
                        onValidate={this.validateScopeName}
                        onClose={this.defineNewScope}
                    />
                    <Button
                        caption="Define New Scope"
                        onClick={this.triggerNewScope}
                    />
                </div>
            </Popup>

        );
    }

    public open = (currentElement: HTMLElement): void => {
        this.popupRef.current?.open(currentElement.getBoundingClientRect());
    };

    /**
     * Called by the scope dialog when a new scope addition is in progress.
     *
     * @param closing True if this validation runs for the entire editing process, otherwise false.
     * @param values The value currently shown in the dialog.
     *
     * @returns An evaluation result for the input values.
     */
    private validateScopeName = (closing: boolean, values: IDialogValues): IDialogValidations => {
        const { currentCustomScopes } = this.state;

        const result: IDialogValidations = { messages: {} };
        const sectionValues = values.sections[0].values;

        let newName = sectionValues.scopeName.value as string || "";
        if (typeof newName === "string") {
            newName = newName.trim();
            if (newName.length === 0) {
                result.messages.scopeName = "The new scope name cannot be empty.";
            } else if (currentCustomScopes.includes(newName)) {
                result.messages.scopeName = "A scope with this name exists already.";
            }
        }

        return result;
    };

    /**
     * Triggered by the user to add a new syntax color scope.
     *
     * @param closure The selected action.
     * @param values The object with new theme name.
     */
    private defineNewScope = (closure: DialogResponseClosure, values: IDialogValues): void => {
        if (closure === DialogResponseClosure.Accept) {
            const { currentCustomScopes } = this.state;
            const sectionValues = values.sections[0].values;

            this.setState({
                currentCustomScopes: [sectionValues.scopeName.value as string, ...currentCustomScopes].sort(),
            });
        }
    };

    private triggerNewScope = (): void => {
        this.newScopeDialogRef.current?.show(
            {
                id: "",
                sections: new Map<string, IDialogSection>([
                    ["", {
                        values: {
                            scopeName: {
                                caption: "Name",
                                value: "",
                            },
                        },
                    }]]),
            },
            { title: "Enter a name for the new scope" },
        );
    };
}
