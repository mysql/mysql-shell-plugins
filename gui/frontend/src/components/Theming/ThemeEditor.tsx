/*
 * Copyright (c) 2020, 2023, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import "./assets/ThemeEditor.css";

import { ComponentChild } from "preact";

import { ThemeManager } from "./ThemeManager";
import { ThemePreview } from "./Preview/ThemePreview";
import { ThemeEditorCore } from "./ThemeEditorCore";
import { ComponentBase } from "../ui/Component/ComponentBase";
import { Container, Orientation } from "../ui/Container/Container";

export class ThemeEditor extends ComponentBase {

    private themeHasChanged = false;
    private changeTimer: ReturnType<typeof setTimeout>;

    public render(): ComponentChild {
        const className = {
            host: this.getEffectiveClassNames(["themeEditorHost"]),
            editor: this.getEffectiveClassNames(["themeEditor"]),
        };

        return (
            <Container
                orientation={Orientation.LeftToRight}
                className={className.host}
            >
                <ThemeEditorCore
                    onThemeChange={this.handleThemeChange}
                />
                <ThemePreview />
            </Container>
        );
    }

    private handleThemeChange = (): void => {
        this.themeHasChanged = true;
        clearTimeout(this.changeTimer);
        this.changeTimer = setTimeout((): void => {
            this.themeHasChanged = false;
            ThemeManager.get.saveTheme();
        }, 1000);

    };
}
