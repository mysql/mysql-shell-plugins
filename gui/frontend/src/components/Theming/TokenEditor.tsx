/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import closeButton from "../../assets/images/close2.svg";
import cloneButton from "../../assets/images/clone.svg";

import React from "react";
import Color from "color";

import {
    Component, IComponentProperties, Container, ITag, CheckState, Orientation, Button, Icon, Grid, GridCell, Label,
    Input, TagInput, Checkbox, ICheckboxProperties, IInputChangeProperties, ColorField, IColorFieldProperties,
} from "../ui";
import { ITokenEntry } from "./ThemeManager";

export interface ITokenEditorProperties extends IComponentProperties {
    token: ITokenEntry;

    onScopeListClick: (e: React.SyntheticEvent) => void;
    onChange: (id: string, token: ITokenEntry) => void;
    onRemove: (id: string, token: ITokenEntry) => void;
    onDuplicate: (id: string, token: ITokenEntry) => void;
}

export class TokenEditor extends Component<ITokenEditorProperties> {

    public constructor(props: ITokenEditorProperties) {
        super(props);

        this.addHandledProperties("onScopeListClick", "onRemove");
    }

    public render(): React.ReactNode {
        const { id, token, onScopeListClick } = this.props;

        const fontStyles = token.settings.fontStyle?.split(" ") || [];
        const name = token.name || "";
        const scopes = Array.isArray(token.scope) ? token.scope : [token.scope];
        const bold = fontStyles.includes("bold") ? CheckState.Checked : CheckState.Unchecked;
        const italic = fontStyles.includes("italic") ? CheckState.Checked : CheckState.Unchecked;
        const underline = fontStyles.includes("underline") ? CheckState.Checked : CheckState.Unchecked;
        const foreground = token.settings.foreground ? new Color(token.settings.foreground) : undefined;
        const background = token.settings.background ? new Color(token.settings.background) : undefined;

        const elementTags: ITag[] = scopes.map((tag: string) => {
            return {
                id: tag,
                caption: tag,
            };
        });

        return (
            <Container
                className="syntaxStyleEntry"
                orientation={Orientation.TopDown}
            >
                <Button
                    round
                    id={id}
                    className="styleEntryDeleteButton"
                    data-tooltip="Delete this syntax entry"
                    onClick={this.handleDeleteTokenClick}
                >
                    <Icon src={closeButton} data-tooltip="inherit" />
                </Button>

                <Button
                    round
                    id={id}
                    className="styleEntryCloneButton"
                    data-tooltip="Clone this syntax entry"
                    onClick={this.handleCloneTokenClick}
                >
                    <Icon src={cloneButton} data-tooltip="inherit" />
                </Button>

                <Grid columns={2} rowGap={4} columnGap={16}>
                    <GridCell>
                        <Label caption="Name:" />
                    </GridCell>
                    <GridCell
                    >
                        <Input
                            id={id}
                            placeholder="<no name assigned>"
                            value={name}
                            onChange={this.handleNameChange}
                        />
                    </GridCell>
                    <GridCell columnSpan={2}>
                        <TagInput
                            id={id}
                            tags={elementTags}
                            orientation={Orientation.LeftToRight}
                            removable
                            className="scopeList"
                            data-tooltip="Click here to add more tags"
                            style={{ flex: "1 1 auto" }}
                            onClick={onScopeListClick}
                            onAdd={this.addScopeToTokenEntry}
                            onRemove={this.removeScopeFromTokenEntry}
                        />
                    </GridCell>
                    <GridCell
                        orientation={Orientation.TopDown}
                        style={{ overflow: "visible" }}
                    >
                        <Container className="syntaxColor">
                            <Label caption="Foreground" style={{ flexGrow: 1 }} />
                            <ColorField
                                id="foreground"
                                initialColor={foreground}
                                onChange={this.handleColorChange}
                            />
                        </Container>
                        <Container className="syntaxColor">
                            <Label caption="Background" style={{ flexGrow: 1 }} />
                            <ColorField id="background" initialColor={background} onChange={this.handleColorChange} />
                        </Container>
                    </GridCell>
                    <GridCell orientation={Orientation.TopDown}>
                        <Checkbox id="bold" caption="bold" checkState={bold} onChange={this.handleCheckChange} />
                        <Checkbox id="italic" caption="italic" checkState={italic} onChange={this.handleCheckChange} />
                        <Checkbox
                            id="underline"
                            caption="underline"
                            checkState={underline}
                            onChange={this.handleCheckChange}
                        />
                    </GridCell>
                </Grid>
            </Container>
        );
    }

    private handleDeleteTokenClick = (): void => {
        const { id, token, onRemove } = this.props;

        onRemove(id!, token);
    };

    private handleCloneTokenClick = (): void => {
        const { id, token, onDuplicate } = this.props;

        onDuplicate(id!, token);
    };

    private addScopeToTokenEntry = (id: string): void => {
        const { onChange, token } = this.props;
        const scopes = Array.isArray(token.scope) ? token.scope : [token.scope];

        token.scope = [...new Set(scopes).add(id)].sort();
        this.forceUpdate();
        onChange(id, token);
    };

    /**
     * Called when the user wants to remove a tag from a token entry.
     *
     * @param id The id of the tag to remove.
     */
    private removeScopeFromTokenEntry = (id: string): void => {
        const { onChange, token } = this.props;
        const scopes = Array.isArray(token.scope) ? token.scope : [token.scope];

        const set = new Set(scopes);
        set.delete(id);
        token.scope = [...set].sort();
        this.forceUpdate();
        onChange(id, token);
    };

    private handleCheckChange = (checkState: CheckState, props: ICheckboxProperties): void => {
        const { onChange, token } = this.props;
        const fontStyles = new Set(token.settings.fontStyle?.split(" ") || []);

        switch (props.id) {
            case "bold": {
                if (checkState === CheckState.Checked) {
                    fontStyles.add("bold");
                } else {
                    fontStyles.delete("bold");
                }

                break;
            }

            case "italic": {
                if (checkState === CheckState.Checked) {
                    fontStyles.add("italic");
                } else {
                    fontStyles.delete("italic");
                }

                break;
            }

            case "underline": {
                if (checkState === CheckState.Checked) {
                    fontStyles.add("underline");
                } else {
                    fontStyles.delete("underline");
                }

                break;
            }

            default: {
                break;
            }
        }

        token.settings.fontStyle = Array.from(fontStyles).join(" ");

        this.forceUpdate();
        onChange(props.id!, token);
    };

    private handleColorChange = (color: Color | undefined, props: IColorFieldProperties): void => {
        const { onChange, token } = this.props;

        const foreground = props.id === "foreground";
        const colorString = color?.hexa();
        if (foreground) {
            token.settings.foreground = colorString;
        } else {
            token.settings.background = colorString;
        }

        this.forceUpdate();
        onChange(props.id!, token);
    };

    private handleNameChange = (e: React.ChangeEvent, props: IInputChangeProperties): void => {
        const { onChange, token } = this.props;
        token.name = props.value;
        this.forceUpdate();

        onChange(props.id!, token);
    };
}
