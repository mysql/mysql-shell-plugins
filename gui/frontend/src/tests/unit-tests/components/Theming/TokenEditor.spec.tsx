/*
 * Copyright (c) 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";
import React from "react";
import { ITokenEditorProperties, TokenEditor } from "../../../../components/Theming/TokenEditor";
import { snapshotFromWrapper } from "../../test-helpers";


describe("Token editor testing", () => {

    it("Token editor properties test", () => {
        const component = mount(
            <TokenEditor
                token={ { name: "token1", scope: "local", settings: {}} }
                id="tokenEditor1"
                onScopeListClick={jest.fn}
                onChange={jest.fn}
                onRemove={jest.fn}
                onDuplicate={jest.fn}
            />,
        );
        expect(component).toBeTruthy();
        const props = component.props() as ITokenEditorProperties;
        expect(props.token).toEqual({ name: "token1", scope: "local", settings: {}});
        component.unmount();
    });

    it("Render test", () => {
        const component = mount(
            <TokenEditor
                token={ { name: "token1", scope: "local", settings: {}} }
                id="tokenEditor1"
                onScopeListClick={jest.fn}
                onChange={jest.fn}
                onRemove={jest.fn}
                onDuplicate={jest.fn}
            />,
        );
        expect (snapshotFromWrapper(component)).toMatchSnapshot();
        component.unmount();
    });

});
