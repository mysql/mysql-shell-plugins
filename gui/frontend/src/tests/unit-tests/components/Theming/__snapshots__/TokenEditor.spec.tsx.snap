// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token editor testing Render test 1`] = `
<TokenEditor
  id="tokenEditor1"
  onChange={[Function]}
  onDuplicate={[Function]}
  onRemove={[Function]}
  onScopeListClick={[Function]}
  token={
    Object {
      "name": "token1",
      "scope": "local",
      "settings": Object {},
    }
  }
>
  <Container
    className="syntaxStyleEntry"
    orientation="column"
  >
    <div
      className="msg container fixedScrollbar syntaxStyleEntry"
      style={
        Object {
          "alignItems": undefined,
          "flexDirection": "column",
          "flexWrap": undefined,
          "justifyContent": undefined,
        }
      }
    >
      <Button
        className="styleEntryDeleteButton"
        data-tooltip="Delete this syntax entry"
        id="tokenEditor1"
        onClick={[Function]}
        round={true}
      >
        <button
          className="msg button round styleEntryDeleteButton"
          data-tooltip="Delete this syntax entry"
          id="tokenEditor1"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          <Icon
            data-tooltip="inherit"
            disabled={false}
            src="close2.svg"
          >
            <div
              className="msg icon"
              data-tooltip="inherit"
              style={
                Object {
                  "WebkitMaskImage": "url(close2.svg)",
                  "backgroundColor": undefined,
                  "height": undefined,
                  "maskImage": "url(close2.svg)",
                  "width": undefined,
                }
              }
            />
          </Icon>
        </button>
      </Button>
      <Button
        className="styleEntryCloneButton"
        data-tooltip="Clone this syntax entry"
        id="tokenEditor1"
        onClick={[Function]}
        round={true}
      >
        <button
          className="msg button round styleEntryCloneButton"
          data-tooltip="Clone this syntax entry"
          id="tokenEditor1"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          <Icon
            data-tooltip="inherit"
            disabled={false}
            src="clone.svg"
          >
            <div
              className="msg icon"
              data-tooltip="inherit"
              style={
                Object {
                  "WebkitMaskImage": "url(clone.svg)",
                  "backgroundColor": undefined,
                  "height": undefined,
                  "maskImage": "url(clone.svg)",
                  "width": undefined,
                }
              }
            />
          </Icon>
        </button>
      </Button>
      <Grid
        columnGap={16}
        columns={2}
        rowGap={4}
      >
        <div
          className="msg grid"
          style={
            Object {
              "columnGap": "16px",
              "gridTemplateColumns": "repeat(2, 1fr)",
              "rowGap": "4px",
            }
          }
        >
          <GridCell>
            <Container
              className="msg gridCell"
              orientation="row"
              style={Object {}}
            >
              <div
                className="msg container fixedScrollbar gridCell"
                style={
                  Object {
                    "alignItems": undefined,
                    "flexDirection": "row",
                    "flexWrap": undefined,
                    "justifyContent": undefined,
                  }
                }
              >
                <Label
                  caption="Name:"
                >
                  <label
                    className="msg label"
                    data-tooltip="expand"
                    style={
                      Object {
                        "textAlign": undefined,
                      }
                    }
                  >
                    Name:
                  </label>
                </Label>
              </div>
            </Container>
          </GridCell>
          <GridCell>
            <Container
              className="msg gridCell"
              orientation="row"
              style={Object {}}
            >
              <div
                className="msg container fixedScrollbar gridCell"
                style={
                  Object {
                    "alignItems": undefined,
                    "flexDirection": "row",
                    "flexWrap": undefined,
                    "justifyContent": undefined,
                  }
                }
              >
                <Input
                  id="tokenEditor1"
                  onChange={[Function]}
                  placeholder="<no name assigned>"
                  spellCheck={true}
                  value="token1"
                >
                  <input
                    className="msg input"
                    id="tokenEditor1"
                    onKeyDown={[Function]}
                    oninput={[Function]}
                    placeholder="<no name assigned>"
                    spellCheck={true}
                    style={
                      Object {
                        "textAlign": undefined,
                      }
                    }
                    type="text"
                    value="token1"
                  />
                </Input>
              </div>
            </Container>
          </GridCell>
          <GridCell
            columnSpan={2}
          >
            <Container
              className="msg gridCell"
              orientation="row"
              style={
                Object {
                  "gridColumn": "span 2",
                }
              }
            >
              <div
                className="msg container fixedScrollbar gridCell"
                style={
                  Object {
                    "alignItems": undefined,
                    "flexDirection": "row",
                    "flexWrap": undefined,
                    "gridColumn": "span 2",
                    "justifyContent": undefined,
                  }
                }
              >
                <TagInput
                  className="scopeList"
                  data-tooltip="Click here to add more tags"
                  id="tokenEditor1"
                  onAdd={[Function]}
                  onClick={[Function]}
                  onRemove={[Function]}
                  orientation="row"
                  removable={true}
                  style={
                    Object {
                      "flex": "1 1 auto",
                    }
                  }
                  tags={
                    Array [
                      Object {
                        "caption": "local",
                        "id": "local",
                      },
                    ]
                  }
                >
                  <Container
                    className="msg tagInput scopeList"
                    data-tooltip="Click here to add more tags"
                    id="tokenEditor1"
                    onClick={[Function]}
                    onDragEnter={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    orientation="row"
                    style={
                      Object {
                        "flex": "1 1 auto",
                      }
                    }
                    wrap="wrap"
                  >
                    <div
                      className="msg container fixedScrollbar tagInput scopeList"
                      data-tooltip="Click here to add more tags"
                      id="tokenEditor1"
                      onClick={[Function]}
                      onDragEnter={[Function]}
                      onDragLeave={[Function]}
                      onDragOver={[Function]}
                      onDragStart={[Function]}
                      onDrop={[Function]}
                      style={
                        Object {
                          "alignItems": undefined,
                          "flex": "1 1 auto",
                          "flexDirection": "row",
                          "flexWrap": "wrap",
                          "justifyContent": undefined,
                        }
                      }
                    >
                      <Label
                        className="tag"
                        draggable={true}
                        key="local"
                      >
                        <label
                          className="msg label tag"
                          data-tooltip="expand"
                          draggable={true}
                          style={
                            Object {
                              "textAlign": undefined,
                            }
                          }
                        >
                          local
                          <Button
                            id="local"
                            onClick={[Function]}
                          >
                            <button
                              className="msg button"
                              id="local"
                              onClick={[Function]}
                              onMouseDown={[Function]}
                              style={
                                Object {
                                  "flexDirection": undefined,
                                }
                              }
                              tabIndex={0}
                            >
                              <Icon
                                as="span"
                                disabled={false}
                                src="close.svg"
                              >
                                <span
                                  className="msg icon"
                                  style={
                                    Object {
                                      "WebkitMaskImage": "url(close.svg)",
                                      "backgroundColor": undefined,
                                      "height": undefined,
                                      "maskImage": "url(close.svg)",
                                      "width": undefined,
                                    }
                                  }
                                />
                              </Icon>
                            </button>
                          </Button>
                        </label>
                      </Label>
                    </div>
                  </Container>
                </TagInput>
              </div>
            </Container>
          </GridCell>
          <GridCell
            orientation="column"
            style={
              Object {
                "overflow": "visible",
              }
            }
          >
            <Container
              className="msg gridCell"
              orientation="column"
              style={
                Object {
                  "overflow": "visible",
                }
              }
            >
              <div
                className="msg container fixedScrollbar gridCell"
                style={
                  Object {
                    "alignItems": undefined,
                    "flexDirection": "column",
                    "flexWrap": undefined,
                    "justifyContent": undefined,
                    "overflow": "visible",
                  }
                }
              >
                <Container
                  className="syntaxColor"
                  orientation="row"
                >
                  <div
                    className="msg container fixedScrollbar syntaxColor"
                    style={
                      Object {
                        "alignItems": undefined,
                        "flexDirection": "row",
                        "flexWrap": undefined,
                        "justifyContent": undefined,
                      }
                    }
                  >
                    <Label
                      caption="Foreground"
                      style={
                        Object {
                          "flexGrow": 1,
                        }
                      }
                    >
                      <label
                        className="msg label"
                        data-tooltip="expand"
                        style={
                          Object {
                            "flexGrow": 1,
                          }
                        }
                      >
                        Foreground
                      </label>
                    </Label>
                    <ColorField
                      id="foreground"
                      onChange={[Function]}
                    >
                      <div
                        className="msg colorField  invalid"
                        draggable={true}
                        id="foreground"
                        onClick={[Function]}
                        onDragEnter={[Function]}
                        onDragLeave={[Function]}
                        onDragOver={[Function]}
                        onDragStart={[Function]}
                        onDrop={[Function]}
                        onKeyDown={[Function]}
                        role="button"
                        tabIndex={0}
                      >
                        <Image
                          className="inner"
                          disabled={false}
                          src="empty-hatch.svg"
                        >
                          <img
                            className="msg image inner"
                            src="empty-hatch.svg"
                          />
                        </Image>
                      </div>
                    </ColorField>
                  </div>
                </Container>
                <Container
                  className="syntaxColor"
                  orientation="row"
                >
                  <div
                    className="msg container fixedScrollbar syntaxColor"
                    style={
                      Object {
                        "alignItems": undefined,
                        "flexDirection": "row",
                        "flexWrap": undefined,
                        "justifyContent": undefined,
                      }
                    }
                  >
                    <Label
                      caption="Background"
                      style={
                        Object {
                          "flexGrow": 1,
                        }
                      }
                    >
                      <label
                        className="msg label"
                        data-tooltip="expand"
                        style={
                          Object {
                            "flexGrow": 1,
                          }
                        }
                      >
                        Background
                      </label>
                    </Label>
                    <ColorField
                      id="background"
                      onChange={[Function]}
                    >
                      <div
                        className="msg colorField  invalid"
                        draggable={true}
                        id="background"
                        onClick={[Function]}
                        onDragEnter={[Function]}
                        onDragLeave={[Function]}
                        onDragOver={[Function]}
                        onDragStart={[Function]}
                        onDrop={[Function]}
                        onKeyDown={[Function]}
                        role="button"
                        tabIndex={0}
                      >
                        <Image
                          className="inner"
                          disabled={false}
                          src="empty-hatch.svg"
                        >
                          <img
                            className="msg image inner"
                            src="empty-hatch.svg"
                          />
                        </Image>
                      </div>
                    </ColorField>
                  </div>
                </Container>
              </div>
            </Container>
          </GridCell>
          <GridCell
            orientation="column"
          >
            <Container
              className="msg gridCell"
              orientation="column"
              style={Object {}}
            >
              <div
                className="msg container fixedScrollbar gridCell"
                style={
                  Object {
                    "alignItems": undefined,
                    "flexDirection": "column",
                    "flexWrap": undefined,
                    "justifyContent": undefined,
                  }
                }
              >
                <Checkbox
                  caption="bold"
                  checkState={0}
                  disabled={false}
                  id="bold"
                  onChange={[Function]}
                >
                  <label
                    className="msg checkbox unchecked"
                    id="bold"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    tabIndex={0}
                  >
                    <span
                      className="checkMark"
                    />
                    bold
                  </label>
                </Checkbox>
                <Checkbox
                  caption="italic"
                  checkState={0}
                  disabled={false}
                  id="italic"
                  onChange={[Function]}
                >
                  <label
                    className="msg checkbox unchecked"
                    id="italic"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    tabIndex={0}
                  >
                    <span
                      className="checkMark"
                    />
                    italic
                  </label>
                </Checkbox>
                <Checkbox
                  caption="underline"
                  checkState={0}
                  disabled={false}
                  id="underline"
                  onChange={[Function]}
                >
                  <label
                    className="msg checkbox unchecked"
                    id="underline"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    tabIndex={0}
                  >
                    <span
                      className="checkMark"
                    />
                    underline
                  </label>
                </Checkbox>
              </div>
            </Container>
          </GridCell>
        </div>
      </Grid>
    </div>
  </Container>
</TokenEditor>
`;
